enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -198,18 +198,39 @@ With the `--external-cloud-volume-plugin` flag set, existing designs for this tr
 
 #### Pre-release / Development Feature Gate
 
-In order to achieve desired cluster configuration for cloud components during development and technical preview, we will add a new [feature gate](https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html) `ExternalCloudProvider`, which will allow developers and customers to force their cluster to upgrade from in-tree to out-of-tree cloud providers.
+In order to achieve desired cluster configuration for cloud components during development and technical preview, we will add a new OpenShift [FeatureGate](https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html) resource named `ExternalCloudProvider`, which will allow developers and customers to force their cluster to upgrade from in-tree to out-of-tree cloud providers.
 
-By using this feature gate with `TechPreviewNoUpgrade` `FeatureSet`, users will prevent their clusters from upgrading to newer versions of OpenShift as with all other feature gates.
+This feature gate will be shipped with `TechPreviewNoUpgrade` `FeatureSet`, preventing upgrades.
 
 The feature gate will be used to determine whether or not operators should migrate their components to the out-of-tree configuration.
-If the feature gate  is listed (via a custom or technical preview feature set), the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for any platform.
+If the feature gate is found, the operators for components such as `kube-apiserver`, `kube-controller-manager` and `kubelet` will update their component's configuration accordingly to allow hosting external `CCM` for current platform.
 
-Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the `TechPreviewNoUpgrade` feature gate and will automatically upgrade as part of a 4.N to 4.N+1 upgrade using default feature set.
+Once an out-of-tree provider implementation is ready to be released (GA), it will no longer depend on the feature gate, allowing production clusters to automatically upgrade to the out-of-tree implementation during 4.N to 4.N+1 upgrade.
 
 Once all providers are moved out-of-tree, the feature gate will be safe to remove.
 
-This feature gate will be commonly used in CI tasks, to ensure upgrade stability.
+This feature gate will be used in CI tasks, to test migration stability.
+
+#### Migration procedure
+
+The process of the migration will be managed by corresponding operators. Communication will be done via conditions on the `ClusterOperator` resource.
+
+The plan for migration is the following:
+
+1) `kubelet` and `kube-controller-manager`.
+    1) Disable cloud loops in it's operands.
+    2) Wait for migration to complete (deployment roll out, kubelet restarted)
```

Comment:
how? Waiting for which signal exactly?

Reply:
This will probably be complex to dive into, but at present, do we know if the KCMO waits until the updated deployment/pods are running/ready before setting it's operator status back to normal? (I assume during a rollout it sets it to progressing but this assumption may not be true). If it does do this, we can piggy back on that. I'm hoping there's also something similar for the kubelet rollouts but we need to check on both of these fronts

Reply:
I suggest we would wait for `Progressing` condition to be set on false. Similarly, backwards migration will watch for `CCCMO` cluster operator to get into `Progressing` false before starting rollback.

Reply:
I think the operators that are doing any rollout/rollback should only care about the `ExternalCloudProvider` condition and not the progressing condition. I think this question is more a case of when will that be added by the operator.

So for KCMO, we want it to add it only when it has updated the deployment for KCM, and the old pods have all stopped, and the new pods are ready. I'm hoping it has logic for this already but we should double check (and for the Kubelet rollouts too, I imagine MCO has some status it sets on the operator while it's rolling out, that is the logic we should be able to piggy back on)